runtime/sema/type_tags.go

```diff
@@ -336,39 +336,58 @@ var (
 	// to be included in the mask without knowing their member types.
 	// Hence, they are checked on demand in `getSuperTypeOfDerivedTypes()`.
 	AnyStructTypeTag = newTypeTagFromLowerMask(anyStructTypeMask).
-				Or(NeverTypeTag).
-				Or(NumberTypeTag).
-				Or(StringTypeTag).
-				Or(ReferenceTypeTag).
-				Or(NilTypeTag).
-				Or(BoolTypeTag).
-				Or(CharacterTypeTag).
-				Or(VoidTypeTag).
-				Or(MetaTypeTag).
-				Or(PathTypeTag).
-				Or(AddressTypeTag).
-				Or(BlockTypeTag).
-				Or(DeployedContractTypeTag).
-				Or(CapabilityTypeTag).
-				Or(FunctionTypeTag)
+		Or(NeverTypeTag).
+		Or(NumberTypeTag).
+		Or(StringTypeTag).
+		Or(ReferenceTypeTag).
+		Or(NilTypeTag).
+		Or(BoolTypeTag).
+		Or(CharacterTypeTag).
+		Or(VoidTypeTag).
+		Or(MetaTypeTag).
+		Or(PathTypeTag).
+		Or(AddressTypeTag).
+		Or(BlockTypeTag).
+		Or(DeployedContractTypeTag).
+		Or(CapabilityTypeTag).
+		Or(FunctionTypeTag)
 
 	AnyResourceTypeTag = newTypeTagFromLowerMask(anyResourceTypeMask)
 
 	AnyTypeTag = newTypeTagFromLowerMask(anyTypeMask).
-			Or(AnyStructTypeTag).
-			Or(AnyResourceTypeTag).
-			Or(ConstantSizedTypeTag).
-			Or(VariableSizedTypeTag).
-			Or(DictionaryTypeTag).
-			Or(GenericTypeTag).
-			Or(InterfaceTypeTag).
-			Or(TransactionTypeTag).
-			Or(RestrictedTypeTag)
+		Or(AnyStructTypeTag).
+		Or(AnyResourceTypeTag).
+		Or(ConstantSizedTypeTag).
+		Or(VariableSizedTypeTag).
+		Or(DictionaryTypeTag).
+		Or(GenericTypeTag).
+		Or(InterfaceTypeTag).
+		Or(TransactionTypeTag).
+		Or(RestrictedTypeTag)
 )
 
 // Methods
 
 func LeastCommonSuperType(types ...Type) Type {
+
+	superType := leastCommonSuperType(types...)
+	if superType == InvalidType {
+		return superType
+	}
+
+	// Do a sanity check to see if all types are
+	// actually subtypes of the inferred supertype.
+	for _, typ := range types {
+		if !IsSubType(typ, superType) {
+			return InvalidType
+		}
```

Comment:
Do we really want this sanity checking running on actual contracts? Or is it possible only to enable this in the language server or linter? Perhaps when `extendedElaboration` is enabled?

Reply:
We need this by default to correct any bugs in the implementation/logic of the super-type inference, if ever there were/will be any. 
e.g: adding a new type can unlock new paths in the logic (like the one we just had), which may have not been thought of before, and as a result, if a wrong type is inferred, then that should be prevented. Otherwise, it could cause type issues at runtime.

Reply:
It sort of ensures that the supertype inference is compliant with (i.e: inverse of) `IsSubType` check